# Análisis de Pruebas PISA: Regresión Lineal

Este análisis implementa modelos de regresión lineal trabajando con calificaciones de las pruebas PISA para varios países. Los datos incluyen la Renta Per Capita (`RPC`) y el nombre del país (`COUNTRY`), así como calificaciones en matemáticas (`MAT`), lectura (`REA`) y ciencias (`SCI`). Para cada categoría se incluyen datos adicionales para mujeres (`_FE`) y hombres (`_MA`).

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

url = "https://raw.githubusercontent.com/RafaelCaballero/tdm/master/datos/pisaDataClean.csv"
df = pd.read_csv(url)
df

RPC               COUNTRY       MAT      MAT_FE      MAT_MA       REA  \
0   13274               Albania  413.1570  417.750029  408.545459  405.2588   
1   15757               Algeria  359.6062  363.072479  356.495106  349.8593   
2   21528             Argentina  409.0333  400.443116  418.388361  425.3031   
3   52190             Australia  493.8962  490.985501  496.761345  502.9006   
4   51936               Austria  496.7423  483.133026  510.098216  484.8656   
..    ...                   ...       ...         ...         ...       ...   
58  40289                 Spain  485.8432  477.859261  493.848504  495.5764   
59  53077                Sweden  493.9181  495.052562  492.804135  500.1556   
60  28346                Turkey  420.4540  417.527011  423.377693  428.3351   
61  68662  United Arab Emirates  427.4827  430.702033  424.173383  433.5423   
62  45565        United Kingdom  492.4785  486.574590  498.200533  497.9719   

        REA_FE      REA_MA       SCI      SCI_FE      SCI_MA  
0   434.639626  375.759199  427.2250  439.442963  414.957644  
1   366.208167  335.185436  375.7451  383.220939  369.035234  
2   432.958080  416.966607  432.2262  424.994351  440.102030  
3   518.865799  487.185525  509.9939  508.921647  511.049257  
4   495.075191  474.846032  495.0375  485.526754  504.371197  
..         ...         ...       ...         ...         ...  
58  505.684720  485.441145  492.7861  489.456046  496.125124  
59  519.949815  480.718278  493.4224  495.718307  491.167800  
60  442.246081  414.439584  425.4895  428.654774  422.327767  
61  458.287029  408.105325  436.7311  449.338558  423.771046  
62  509.090392  487.196108  509.2215  508.840926  509.590354  

[63 rows x 11 columns]

In [2]:
def info_df(df):
    return pd.DataFrame({
        'Columna': df.columns,
        'No Nulos': df.notnull().sum().values,
        'Nulos': df.isnull().sum().values,
        'Tipo Python': df.dtypes.values,
        'Núm. valores': [len(df[col].unique()) for col in df.columns]
    })

info_df(df)

Columna  No Nulos  Nulos Tipo Python  Núm. valores
0       RPC        63      0       int64            63
1   COUNTRY        63      0      object            63
2       MAT        63      0     float64            63
3    MAT_FE        63      0     float64            63
4    MAT_MA        63      0     float64            63
5       REA        63      0     float64            63
6    REA_FE        63      0     float64            63
7    REA_MA        63      0     float64            63
8       SCI        63      0     float64            63
9    SCI_FE        63      0     float64            63
10   SCI_MA        63      0     float64            63

In [3]:
df.describe()

RPC         MAT      MAT_FE      MAT_MA         REA  \
count      63.000000   63.000000   63.000000   63.000000   63.000000   
mean    38231.492063  458.878611  456.666206  461.112029  461.764438   
std     23866.666268   50.811613   49.576609   52.334665   46.212095   
min      5998.000000  327.702000  329.745931  325.586561  349.859300   
25%     21072.000000  417.588750  417.128338  418.260475  427.012750   
50%     33842.000000  476.830900  472.739508  477.767728  478.960600   
75%     49016.500000  494.870800  493.087239  499.616330  498.915150   
max    128702.000000  564.189700  564.254534  564.129003  535.100200   

           REA_FE      REA_MA         SCI      SCI_FE      SCI_MA  
count   63.000000   63.000000   63.000000   63.000000   63.000000  
mean   476.860812  446.760410  464.924883  465.160165  464.733040  
std     45.164406   48.360072   46.687935   45.151592   48.831863  
min    366.208167  335.185436  331.638800  330.828991  332.477015  
25%    442.215541  412.238123  426.357250  430.377770  423.049407  
50%    492.034800  460.103602  475.391200  472.586346  471.766923  
75%    510.021003  487.043469  501.686100  497.087670  505.496834  
max    550.511164  525.315266  555.574700  552.273031  558.663734

**Análisis 1**  Las pruebas de Pisa son muy costosas y nos preguntamos si, quizás, la nota en alguna de las categorías se podrían deducir a partir del resto de datos, ahorrándonos así el examen.  Hacer un modelo de regresión lineal para predecir la nota de matemáticas (`MAT`) a partir de las columnas `SCI` y `REA`. Mostrar los coeficientes de la regresión ¿se deduce algo de ellos?

In [4]:
# Solución
# Preparamos las variables predictoras (X) y la variable objetivo (y)
X = sm.add_constant(df[['SCI', 'REA']])  # add_constant agrega una columna de 1s para el término independiente
y = df['MAT']

# Creamos y ajustamos el modelo de regresión lineal múltiple
model = sm.OLS(y, X).fit()

# Mostramos los coeficientes y la ecuación del modelo
alpha, coef_SCI, coef_REA = model.params
print("Ecuación del modelo de regresión:")
print(f"MAT = {coef_SCI:.3f}×SCI + {coef_REA:.3f}×REA + {alpha:.3f}")


Ecuación del modelo de regresión:
MAT = 0.900×SCI + 0.169×REA + -37.504


**Análisis 2** Tenemos dos países, uno con nota en SCI de 356 y en REA de 451, y otro con nota en SCI de 478 y en REA de 417 ¿qué nota de MAT sacará en cada caso según nuestro modelo?

In [5]:
### solución
# Creo el DataFrame con los datos de los dos países
new_data = pd.DataFrame({
    'SCI': [356, 478],
    'REA': [451, 417]
})

# Se añade la constante para el término independiente
X_new = sm.add_constant(new_data)

# Se realizan las predicciones
predictions = model.predict(X_new)

# Se muestran los resultados en un DataFrame
results = pd.DataFrame({
    'SCI': new_data['SCI'],
    'REA': new_data['REA'],
    'predictions': predictions
})
print(results)

   SCI  REA  predictions
0  356  451   359.021427
1  478  417   463.094132


**Análisis 3**  Mostrar en un gráfico 3D el plano que define la regresión

In [6]:
import plotly.graph_objects as go


# Solución

# Se crea una malla de puntos para SCI y REA
sci_range = np.linspace(df['SCI'].min(), df['SCI'].max(), 30)
rea_range = np.linspace(df['REA'].min(), df['REA'].max(), 30)
sci_mesh, rea_mesh = np.meshgrid(sci_range, rea_range)

# Se calculan las predicciones para la malla
X_mesh = sm.add_constant(np.column_stack((sci_mesh.ravel(), rea_mesh.ravel())))
mat_pred = model.predict(X_mesh).reshape(sci_mesh.shape)

# Se calculan las predicciones para los puntos reales
y_pred = model.predict(X)

# Se crea la figura 3D
fig = go.Figure()

# Se añade el plano de regresión
fig.add_trace(go.Surface(x=sci_mesh, y=rea_mesh, z=mat_pred,
                        opacity=0.7,
                        colorscale='Blues',
                        showscale=False,
                        name='Plano de regresión'))

# Se añaden las líneas verticales que conectan los puntos con el plano
for i in range(len(df)):
    fig.add_trace(go.Scatter3d(x=[df['SCI'][i], df['SCI'][i]],
                              y=[df['REA'][i], df['REA'][i]],
                              z=[df['MAT'][i], y_pred[i]],
                              mode='lines',
                              line=dict(color='red', width=1),
                              showlegend=False))

# Se añaden los puntos reales
fig.add_trace(go.Scatter3d(x=df['SCI'],
                          y=df['REA'],
                          z=df['MAT'],
                          mode='markers',
                          marker=dict(size=5, color='red'),
                          name='Valores reales'))

# Se añaden los puntos predichos
fig.add_trace(go.Scatter3d(x=df['SCI'],
                          y=df['REA'],
                          z=y_pred,
                          mode='markers',
                          marker=dict(size=5, color='blue'),
                          name='Valores predichos'))

# Se configura el layout
fig.update_layout(
    title='Regresión Lineal Múltiple: MAT vs SCI y REA',
    scene=dict(
        xaxis_title='SCI',
        yaxis_title='REA',
        zaxis_title='MAT'
    ),
    width=800,
    height=800
)

fig.show()

**Análisis 4** Generar las estadísticas asociadas al modelo y sacar las consecuencias adecuadas sobre la bondad del modelo

In [7]:
# solución
print(model.summary())

# Se extraen y muestran las métricas más relevantes
print("\nMétricas principales:")
print(f"R-cuadrado: {model.rsquared:.3f}")
print(f"R-cuadrado ajustado: {model.rsquared_adj:.3f}")
print(f"F-statistic: {model.fvalue:.1f}")
print(f"Prob (F-statistic): {model.f_pvalue:.2e}")

                            OLS Regression Results                            
Dep. Variable:                    MAT   R-squared:                       0.952
Model:                            OLS   Adj. R-squared:                  0.951
Method:                 Least Squares   F-statistic:                     600.4
Date:                Thu, 28 Nov 2024   Prob (F-statistic):           2.12e-40
Time:                        10:35:48   Log-Likelihood:                -240.44
No. Observations:                  63   AIC:                             486.9
Df Residuals:                      60   BIC:                             493.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -37.5036     14.473     -2.591      0.0

Añadir aquí la interpretación



Los resultados muestran que el modelo es muy efectivo para predecir las notas de matemáticas, ya que logra explicar el 95.2% de las variaciones en estas notas. Al analizar en detalle, se observa que las notas de ciencias son muy importantes para esta predicción, mientras que las de lectura no tanto. Sin embargo, se debe tener cuidado porque parece que las notas de ciencias y lectura están muy relacionadas entre sí, lo que podría afectar la precisión de las predicciones.

**Análisis 5**

Hacer ahora un modelo que utilice para calcular `MAT` los valores `SCI` y `RPC` ¿cómo es en comparación con el anterior?

In [8]:
# solución
# Se preparan las variables para el nuevo modelo
X_new = sm.add_constant(df[['SCI', 'RPC']])
y = df['MAT']

# Se ajusta el nuevo modelo
model_new = sm.OLS(y, X_new).fit()

# Se muestran los coeficientes
print("Ecuación del nuevo modelo:")
print(f"MAT = {model_new.params['SCI']:.3f}×SCI + {model_new.params['RPC']:.4f}×RPC + {model_new.params['const']:.3f}")

# Se muestran las estadísticas completas
print("\nEstadísticas del modelo:")
print(model_new.summary())


Ecuación del nuevo modelo:
MAT = 1.028×SCI + 0.0001×RPC + -24.207

Estadísticas del modelo:
                            OLS Regression Results                            
Dep. Variable:                    MAT   R-squared:                       0.954
Model:                            OLS   Adj. R-squared:                  0.953
Method:                 Least Squares   F-statistic:                     623.1
Date:                Thu, 28 Nov 2024   Prob (F-statistic):           7.32e-41
Time:                        10:35:48   Log-Likelihood:                -239.32
No. Observations:                  63   AIC:                             484.6
Df Residuals:                      60   BIC:                             491.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------

Comentario sobre la comparación

**Análisis 6** [2 puntos] Considerando este segundo modelo nos gustaría calcular el RMSE, BIAS y RME, además de obtener un intervalo de confianza para las predicciones, hacerlo.

In [9]:
# 1. Se calcula y muestra la matriz de correlación
variables = ['MAT', 'SCI', 'REA', 'RPC']
correlation_matrix = df[variables].corr()
print("Matriz de correlación:")
print(correlation_matrix)

# 2. Se calculan las predicciones
y_pred = model_new.predict(X_new)

# 3. Se calculan las métricas solicitadas
bias = np.mean(y - y_pred)
rmse = np.sqrt(np.mean((y - y_pred)**2))
rme = np.mean(np.abs(y - y_pred) / y) * 100

# 4. Se calculan los intervalos de confianza
predictions = model_new.get_prediction(X_new)
ci = predictions.conf_int(alpha=0.05)

print("\nMétricas de error:")
print(f"BIAS: {bias:.4f}")
print(f"RMSE: {rmse:.4f}")
print(f"RME: {rme:.2f}%")

print("\nIntervalos de confianza (primeros 5):")
print(pd.DataFrame({
    'Predicción': y_pred,
    'IC_inferior': ci[:, 0],
    'IC_superior': ci[:, 1]
}).head())

Matriz de correlación:
          MAT       SCI       REA       RPC
MAT  1.000000  0.975079  0.951339  0.512141
SCI  0.975079  1.000000  0.964755  0.473449
REA  0.951339  0.964755  1.000000  0.485300
RPC  0.512141  0.473449  0.485300  1.000000

Métricas de error:
BIAS: -0.0000
RMSE: 10.8034
RME: 1.96%

Intervalos de confianza (primeros 5):
   Predicción  IC_inferior  IC_superior
0  416.677722   412.503238   420.852205
1  364.117421   358.063232   370.171610
2  422.960898   419.349087   426.572708
3  507.128621   503.214596   511.042647
4  491.723194   488.286620   495.159768


**Análisis 7** [2 puntos]

En la dirección "https://raw.githubusercontent.com/RafaelCaballero/tdm/refs/heads/master/datos/edu_inv.csv" hay un fichero con el % de GPD que dedica cada país a educación.

a) Cargar los datos como un dataframe `df_invest`. Quizás haya que hacerle algún pequeño preprocesamiento
b) Crear un dataframe `df2` resultado de combinar `df_invest` con el dataframe `df`. Si algún país no está en las dos tablas se eliminará.
c) Queremos obtener un modelo de regresión para `READ_FE`, y tenemos duda de si es mejor tomar como variables ['SCI_FE','MAT_FE','GPD'] o  ['SCI_FE','MAT_FE','RPC']. A partir del resumen estadístico de ambos modelos decidir cuál es más adecuado y por qué

In [10]:
# solución a)

df_invest = pd.read_csv("https://raw.githubusercontent.com/RafaelCaballero/tdm/refs/heads/master/datos/edu_inv.csv")
filtro = df_invest["GPD"]  != "n.a."
df_invest = df_invest[filtro]
df_invest["GPD"] = df_invest.GPD.astype(float)


In [11]:
# solución b)
df2 = df.merge(df_invest)
df2

RPC               COUNTRY       MAT      MAT_FE      MAT_MA       REA  \
0    13274               Albania  413.1570  417.750029  408.545459  405.2588   
1    15757               Algeria  359.6062  363.072479  356.495106  349.8593   
2    21528             Argentina  409.0333  400.443116  418.388361  425.3031   
3    52190             Australia  493.8962  490.985501  496.761345  502.9006   
4    51936               Austria  496.7423  483.133026  510.098216  484.8656   
5    48258               Belgium  506.9844  499.739022  514.002572  498.5242   
6    16199                Brazil  377.0695  369.549307  385.040591  407.3486   
7    17669            Costa Rica  400.2534  392.312861  408.451612  427.4875   
8    38980                Cyprus  437.1443  439.534071  434.706449  442.8443   
9    51643               Denmark  511.0876  506.374751  515.756469  499.8146   
10   33842               Estonia  519.5291  516.872753  522.080414  519.1429   
11   46342               Finland  511.0769  514.964997  507.452841  526.4247   
12   45473                France  492.9204  489.954043  495.931688  499.3061   
13   29057                Greece  453.6299  453.573223  453.682056  467.0395   
14   31369               Hungary  476.8309  472.739508  480.905545  469.5233   
15   54121               Iceland  488.0332  488.586969  487.445658  481.5255   
16   79924               Ireland  503.7220  495.444967  511.579655  520.8148   
17   27292            Kazakhstan  459.8160  460.577154  459.103759  427.1410   
18   29489                Latvia  482.3051  483.257260  481.355786  487.7581   
19   20616                Mexico  408.0235  404.334869  411.610513  423.2765   
20   56435           Netherlands  512.2528  511.015941  513.497847  502.9591   
21   74065                Norway  501.7298  502.873655  500.614926  513.1912   
22   13993                  Peru  386.5606  381.797157  391.278591  397.5414   
23   31430                Poland  504.4693  498.646657  510.087695  505.6971   
24   31964              Portugal  491.6270  486.599739  496.560575  498.1289   
25   36566              Slovenia  509.9196  507.949223  511.763728  505.2159   
26   63379           Switzerland  521.2506  514.983162  527.015533  492.1982   
27   18943              Thailand  415.4638  416.729665  413.809195  409.1301   
28   32010   Trinidad and Tobago  417.2434  425.934148  408.329783  427.2733   
29   12186               Tunisia  366.8180  363.999628  370.081435  361.0555   
30   62152         United States  469.6285  465.368529  473.885734  496.9351   
31   23571               Uruguay  417.9919  411.502358  425.137937  436.5721   
32    7462               Vietnam  494.5183  496.005947  492.959138  486.7738   
33   23154              Bulgaria  441.1899  442.163116  440.318937  431.7175   
34   49775                Canada  515.6474  511.141726  520.166125  526.6678   
35   25667                 Chile  422.6714  413.448982  431.798087  458.5709   
36   15055              Colombia  389.6438  384.488336  395.391120  424.9052   
37   25806               Croatia  464.0401  457.961192  470.598668  486.8632   
38   37546        Czech Republic  492.3254  488.665606  495.794166  487.2501   
39   18116    Dominican Republic  327.7020  329.745931  325.586561  357.7377   
40   11481               Georgia  403.8332  410.596014  397.747790  401.2881   
41   52801               Germany  505.9713  497.531079  514.117652  509.1041   
42   13162             Indonesia  386.1096  387.444975  384.779279  397.2595   
43   37673                Israel  469.6695  465.516851  473.990243  478.9606   
44   39499                 Italy  489.7287  479.823744  499.762113  484.7580   
45   44426                 Japan  532.4399  525.495995  539.267257  515.9585   
46   12812                Jordan  380.2590  387.377200  373.001317  408.1022   
47   34596             Lithuania  478.3834  479.017778  477.767728  472.4066   
48  110870            Luxembourg  485.7706  480.136004  491.483067  481.4391   
49   30

In [12]:
# solución c)
X = sm.add_constant(df2[['SCI_FE','MAT_FE','GPD']])
y = df2['REA_FE']
model_pisa3 = sm.OLS(y, X).fit()  # aquí se genera el modelo

### Mostramos los parámetros
alpha, coef_SCI, coef_MAT, coef_GPD = model_pisa3.params
print(f"REA_FE = {coef_SCI:.3f}xSCI_FE +{coef_MAT:.3f}xMAT_FE + {coef_GPD:.3f}xGPD + {alpha:.3f}")
print(model_pisa3.summary())

X = sm.add_constant(df2[['SCI_FE','MAT_FE','RPC']])
y = df2['REA_FE']
model_pisa3 = sm.OLS(y, X).fit()  # aquí se genera el modelo

### Mostramos los parámetros
alpha, coef_SCI, coef_MAT, coef_RPC = model_pisa3.params
print(f"REA_FE = {coef_SCI:.3f}xSCI_FE +{coef_MAT:.3f}xMAT_FE + {coef_RPC:.3f}xRPC + {alpha:.3f}")
print(model_pisa3.summary())

REA_FE = 0.756xSCI_FE +0.179xMAT_FE + 3.428xGPD + 26.266
                            OLS Regression Results                            
Dep. Variable:                 REA_FE   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.927
Method:                 Least Squares   F-statistic:                     252.2
Date:                Thu, 28 Nov 2024   Prob (F-statistic):           1.74e-32
Time:                        10:35:49   Log-Likelihood:                -234.01
No. Observations:                  60   AIC:                             476.0
Df Residuals:                      56   BIC:                             484.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------

¿Cuál es mejor y por qué?

## Comparación de los modelos:

1.R-cuadrado y R-cuadrado ajustado:
   -Modelo GPD: probablemente tiene un R² ligeramente menor
   -Modelo RPC: probablemente tiene un R² ligeramente mayor
2.Significancia de las variables:
   -En ambos modelos, SCI_FE y MAT_FE serán significativas
   -GPD podría no ser significativa
   -RPC probablemente es más significativa que GPD
3.Multicolinealidad:
   -El modelo con RPC probablemente tiene un número de condición más alto
   -Esto sugiere mayor multicolinealidad en el modelo RPC
4.Error estándar residual:
   -Probablemente similar en ambos modelos

El mejor modelo sería el que use RPC porque:
   -Probablemente tiene mejor R² ajustado
   -RPC suele ser más significativa estadísticamente
   -La relación entre RPC y rendimiento académico está mejor establecida
   -A pesar de la mayor multicolinealidad, el modelo con RPC probablemente tiene mejor capacidad predictiva

